File: 백준/Gold/20055. 컨베이어 벨트 위의 로봇/컨베이어 벨트 위의 로봇.cc
#include <bits/stdc++.h>

using namespace std;
#define FASTIO ios::sync_with_stdio(false); cin.tie(0); cout.tie(0)
#define LL long long
#define ULL unsigned long long

// const int INF = 1987654321;
const int INF = 1e9;
const int MOD = 1e6;
const int MAX = 40001;

struct Belt {
	bool robotExist = false;
	int A;
};

int n, k;
Belt belt[201];
vector<int> robotLoc;
int deadCnt = 0;

void rotate() {
	Belt tmp = belt[2 * n - 1];
	for(int i = 2 * n - 1; i > 0; i--)
		belt[i] = belt[i - 1];
	belt[0] = tmp;

	for(int i = deadCnt; i < robotLoc.size(); i++) {
		robotLoc[i] = (robotLoc[i] + 1) % (2 * n);
	}
}

void moveRobot() {
	for(int rNum = deadCnt; rNum < robotLoc.size(); rNum++) {
		int cLoc = robotLoc[rNum];
		int nLoc = (cLoc + 1) % (2 * n);
		if(belt[nLoc].robotExist || belt[nLoc].A == 0) continue;
		belt[cLoc].robotExist = false;
		belt[nLoc].robotExist = true;
		belt[nLoc].A--;
		robotLoc[rNum] = nLoc;
	}
}

void setRobot() {
	if(belt[0].A > 0) {
		belt[0].robotExist = true;
		belt[0].A--;

		robotLoc.push_back(0);
	}
}

void getRobot() {
	if(belt[n - 1].robotExist) {
		belt[n - 1].robotExist = false;
		deadCnt++;
	}
}

bool isFinish() {
	int cnt = 0;
	for(int i = 0; i < 2 * n; i++) {
		if(belt[i].A <= 0) {
			cnt++;
		}
	}
	return cnt >= k;
}

int solve() {
	int cnt = 1;
	while(true) {
		rotate();
		getRobot();

		moveRobot();
		getRobot();

		setRobot();
		
		if(isFinish()) break;
		cnt++;
	}
	return cnt;
}

int main(void) {
    FASTIO;

	cin >> n >> k;
	for(int i = 0; i < 2 * n; i++) {
		int a; cin >> a;
		belt[i] = {false, a};
	}
	cout << solve() << '\n';

    return 0;
}

File: 백준/Gold/20055. 컨베이어 벨트 위의 로봇/컨베이어 벨트 위의 로봇.java
import java.util.*;
import java.util.stream.Collector;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import java.io.*;

public class Main {

    static int n, k;
    static int[] A = new int[201];
    static int[] robot = new int[201];
    static int cnt = 0;

    static void rotate() {
        int tmp = A[2 * n - 1];
        for(int i = 2 * n - 1; i > 0; i--) {
            A[i] = A[i - 1];
            robot[i] = robot[i - 1];
        }
        A[0] = tmp;
        robot[0] = 0;
    }

    static void move() {
        for(int i = n - 2; i >= 0; i--) {
            if(robot[i] != 1) continue;
            if(robot[i + 1] != 1 && A[i + 1] > 0) {
                robot[i + 1] = 1;
                robot[i] = 0;
                A[i + 1]--;
            }
        }
    }
    static void getRobot() {
        if(robot[n - 1] == 1) robot[n - 1] = 0;
    }

    static void setRobot() {
        if(A[0] > 0) {
            robot[0] = 1;
            A[0]--;
        }
    }

    static void updateK() {
        int sum = 0;
        for(int i = 0; i < 2 * n; i++) {
            if(A[i] <= 0) sum++;
        }
        cnt = sum;
    }

    static int solve() {
        int t = 0;
        while(true) {
            t++;
            rotate();
            getRobot();
            move();
            getRobot();
            setRobot();
            updateK();
            if(cnt >= k) break;
        }
        return t;
    }

    public static void main(String[] args) throws Exception {
        Scanner sc = new Scanner(System.in);

        n = sc.nextInt();
        k = sc.nextInt();
        for(int i = 0; i < 2 * n; i++) {
            A[i] = sc.nextInt();
        }

        System.out.println(solve());
    }
}
